src/LanguageServer/Impl/Documentation/DocstringConverter.cs

```diff
@@ -0,0 +1,460 @@
+
+        private int _blockIndent = 0;
+        private bool CurrentLineIsOutsideBlock => CurrentIndent < _blockIndent;
+        private string CurrentLineWithinBlock => CurrentLine.Substring(_blockIndent);
+
+        private DocstringConverter(string input) {
+            _state = ParseText;
+            _lines = SplitDocstring(input);
+        }
+
+        private string Convert() {
+            while (CurrentLine != null) {
+                var shouldAdvance = _state();
+                if (shouldAdvance) {
+                    _lineNum++;
+                }
+            }
+
+            // Close out any outstanding code blocks.
+            if (_state == ParseBacktickBlock || _state == ParseDoctest || _state == ParseLiteralBlock) {
+                TrimOutputAndAppendLine("```");
+            } else if (_insideInlineCode) {
+                TrimOutputAndAppendLine("`", true);
+            }
+
+            return _builder.ToString().Trim();
+        }
+
+        private void PushAndSetState(Func<bool> next) {
+            if (_state == ParseText) {
+                _insideInlineCode = false;
+            }
+
+            _stateStack.Push(_state);
+            _state = next;
+        }
+
+        private void PopState() {
+            _state = _stateStack.Pop();
+
+            if (_state == ParseText) {
+                // Terminate inline code when leaving a block.
+                _insideInlineCode = false;
+            }
+        }
+
+        private bool ParseText() {
+            if (string.IsNullOrWhiteSpace(CurrentLine)) {
+                _state = ParseEmpty;
+                return false;
+            }
+
+            if (CurrentLine.StartsWith("```")) {
+                AppendLine(CurrentLine);
+                PushAndSetState(ParseBacktickBlock);
+                return true;
+            }
+
+            if (BeginLiteralBlock()) {
+                return false;
+            }
+
+            if (BeginDoctest()) {
+                return true;
+            }
+
+            if (BeginDirective()) {
+                return false;
+            }
+
+            // TODO: Push into Google/Numpy style list parser.
+
+            AppendTextLine(CurrentLine);
+            return true;
+        }
+
+        private void AppendTextLine(string line) {
+            line = PreprocessTextLine(line);
+
+            // Hack: attempt to put directives lines into their own paragraphs.
+            // This should be removed once proper list-like parsing is written.
+            if (!_insideInlineCode && Regex.IsMatch(line, @"^\s*:(param|arg|type|return|rtype|raise|except|var|ivar|cvar|copyright|license)")) {
+                AppendLine();
+            }
+
+            var parts = line.Split('`');
+
+            for (var i = 0; i < parts.Length; i++) {
+                var part = parts[i];
+
+                if (i > 0) {
+                    _insideInlineCode = !_insideInlineCode;
+                    Append("`");
```

Comment:
Can be Append(‘’) ie char

Reply:
Sure, I can add an overload for that below (though this converter is not exactly the most efficient given the large number of replacement, regex matches, and string splits/joins...).